HTTPRequestAsync operator.

leongor opened the pull request:
Related issue is here: https://github.com/IBMStreams/streamsx.inet/issues/306
Related wiki is here: https://github.com/IBMStreams/streamsx.inet/wiki/HTTPRequestAsync-operator

chanskw commented:
Is the Boost library included as part of this pull request?  Do we have to include the Boost library into the toolkit?  I am not sure if we have the rights to distribute Boost library.

leongor commented:
To build the shared library Boost is required, but for operator
boost_system and boost_thread static libs are only needed.

ср, 6 сент. 2017 г. в 20:56, Samantha Chan <notifications@github.com>:

> Is the Boost library included as part of this pull request? Do we have to
> include the Boost library into the toolkit? I am not sure if we have the
> rights to distribute Boost library.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/IBMStreams/streamsx.inet/pull/310#issuecomment-327563983>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvlA2MXzztlwDP0QOrWv_0q3S8V2-3zks5sftzSgaJpZM4POoJY>
> .
>

leongor commented:
Removed the dependency for boost_system and boost_thread static libs.
If you mean the boost::network in include folder - it's an actual cpp-netlib library, like described in this issue: https://github.com/IBMStreams/streamsx.inet/issues/306

joergboe commented:
Are there alternatives to the cpp-netlib library?
There seems to be a depencency to c++11, which is not available on a standard rhel 6 client.

leongor commented:
@joergboe I compiled it for both RHEL 6/7. The c++11 core code compiled separately to a shared library (encapsulating all dependencies inside), so for the SPL developer just 2 operator releases can be provided. If someone wants to build it from the scratch, it'll be documented how to do it on both platform versions.

chanskw commented:
Since this pull request still includes the boost library... I need to review this change before it can be merged.  I want to make sure that there is no licensing issue including them into the repository.

chanskw commented:
Also, is there a reason why you can't use the Boost library shipped with Streams and have to ship your own?

leongor commented:
@chanskw as I already mentioned above, it's not a boost library, it's cpp-netlib library. It has a boost namespace because it's a future candidate to become a part of the boost.

joergboe commented:
I do not like the idea to include a special library in the toolkit  which requires a special dev-toolset on platform rhel6 to compile the lib.
To release the lib with the toolkit requires the go through the open source process.
So again the question is there alternative solution possible?

chanskw commented:
Can you send me a link as to where this library come from?

chanskw commented:
Because we are including additional open-source resources... it maybe that we have to go through OSSC as they are not covered by our Apache license.  I need to take a look before we decide.  But we cannot incorporate without checking.

leongor commented:
@chanskw Here the library: [cpp-netlib](http://cpp-netlib.org)
@joergboe I understand your concern. I didn't find better choice for HTTP Async library in c++. Maybe in java.

chanskw commented:
"@joergboe I compiled it for both RHEL 6/7. The c++11 core code compiled separately to a shared library (encapsulating all dependencies inside), so for the SPL developer just 2 operator releases can be provided. If someone wants to build it from the scratch, it'll be documented how to do it on both platform versions."

@leongor are you also distributing this library?  that we will be distributing c++11 core code as part of the toolkit?

leongor commented:
@chanskw Do you mean as source? Yes, at least library headers should be part of the toolkit.

chanskw commented:
I am putting a -1 for this pull request.  Please seek OSSC approval before contributing again.  Thanks!

joergboe commented:
I try to understand the operator code but I wonder where is the operator xml. file?

leongor commented:
@joergboe Investigating your question I noticed this rule in .gitignore:
/com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/*/*.xml
It filters auto generated Java operator models, but now it also filtered HTTPRequestAsync operator model.
How do you suggest to handle it?

joergboe commented:
We should express the exemption from the rule with !
similar to the native.function/function.xml

leongor commented:
@joergboe Thanks, now it's in.

joergboe commented:
Which version of boost lib is used to compile the toolkit code?

leongor commented:
cpp-netlib was compiled with static boost 1.64, so the toolkit itself has no dependencies on boost.

leongor commented:
I took a basic centos 6 docker image and build the library with the following steps:
```
#!/usr/bin/env bash

# Prepare and install devtoolset-4-gcc-c++
sudo yum install centos-release-scl -y
sudo yum install cmake openssl-devel unzip wget devtoolset-4-gcc-c++ -y

# Download and extract boost 1.64
wget https://dl.bintray.com/boostorg/release/1.64.0/source/boost_1_64_0.tar.gz
tar zxvf boost_1_64_0.tar.gz

# Init and build boost static with c++11
pushd boost_1_64_0 && BOOST_HOME=`pwd`
scl enable devtoolset-4 "./bootstrap.sh"
scl enable devtoolset-4 "./b2 -j8 toolset=gcc variant=release link=static threading=multi address-model=64 cxxflags='-std=c++11 -fPIC'"
popd

# Download and extract cpp-netlib 0.13
wget https://github.com/cpp-netlib/cpp-netlib/archive/0.13-release.zip
unzip 0.13-release.zip

# Config and build cpp-netlib
pushd cpp-netlib-0.13-release
mkdir build && cd build

scl enable devtoolset-4 "cmake -DBOOST_ROOT=${BOOST_HOME} \
                               -DCPP-NETLIB_BUILD_SHARED_LIBS=ON \
                               -DCPP-NETLIB_BUILD_EXAMPLES=OFF \
                               -DCPP-NETLIB_BUILD_TESTS=OFF \
                               -DCPP-NETLIB_STATIC_BOOST=ON \
                               -DCMAKE_BUILD_TYPE=Release \
                               -DCMAKE_POSITION_INDEPENDENT_CODE=ON \
                               -DCMAKE_SHARED_LIBRARY_SONAME_CXX_FLAG='-U' \
                               -DCMAKE_SHARED_LINKER_FLAGS='-Wl,--whole-archive -lboost_system -lboost_thread -Wl,--no-whole-archive' .."
make -j8
popd
```

joergboe commented:
Question relating the operator design:
Is there really the requirement to change the http-method on the fly?
I wonder that the expression-mode for the method-parameter is 'Expression' but the expression-mode for the contentType-parameter is 'CustomLiteral' which is compile time constant.

joergboe commented:
I wonder why the workerQueues parameter has expression-mode 'Expression' but the operator code evaluates the value only at start up. This should be a 'AttributeFree' expression-mode.
Or was it intended to change the workerQueues value on the fly? But this would require an different implementation.

joergboe commented:
I do not see any reason why was operator is not implemented based on the libcurl. The libcurl is available on all relevant platforms and does not require c++11.
Linking of  C++98 objects with C++11 objects is potential dangerous:
[AbiCompatibility](https://gcc.gnu.org/wiki/Cxx11AbiCompatibility)
As long as we do not have integral c++11 support I would not recommend to use c++11 code without urgent need.

leongor commented:
@joergboe 

> Is there really the requirement to change the http-method on the fly?

Yes, the requirement comes from Streaming Pipelines team.

> I wonder that the expression-mode for the method-parameter is 'Expression' but the expression-mode for the contentType-parameter is 'CustomLiteral' which is compile time constant.

'contentType' parameter is a compile time constant indeed, but it's provided for convenience and may be overriden by 'headers' parameter (as one of key/value pairs) which is dynamic.

> I wonder why the workerQueues parameter has expression-mode 'Expression' but the operator code evaluates the value only at start up. This should be a 'AttributeFree' expression-mode.

Yes, that's right.

> I do not see any reason why was operator is not implemented based on the libcurl. The libcurl is available on all relevant platforms and does not require c++11.

libcurl lacks asynchronous non-blocking communication support. 

>  Linking of C++98 objects with C++11 objects is potential dangerous.

Potentially yes, but Streams allows to customize build with '--c++std=c++11' flag, while Streams runtime is still compiled with C++03.

joergboe commented:
The make file does not do what we expect:
Compiling 2 src/httpFunctions.cpp ...
g++ -O3 -DNDEBUG -fPIC -D_REENTRANT -fPIC -D_IBM_STREAMS_VER_=4 -D_IBM_STREAMS_REL_=2 -D_IBM_STREAMS_MOD_=3 -D_IBM_STREAMS_FIX_=0 -isystem/opt/ibm/InfoSphere_Streams/4.2.3.0/include -isystem/opt/ibm/InfoSphere_Streams/4.2.3.0/system/impl/include -isystem/opt/ibm/InfoSphere_Streams/4.2.3.0/ext/include -isystem/opt/ibm/InfoSphere_Streams/4.2.3.0/ext/include/decNumber   -std=gnu++11 -Werror -Wall -Wno-deprecated -I ../opt/downloaded/boost_1_64_0  -c src/httpFunctions.cpp -o cpp/bin/httpFunctions.o -I ./include
src/httpFunctions.cpp: In function ‘curl_slist* com_ibm_streamsx_inet_http::getSList(SPL::list<SPL::rstring>)’:
src/httpFunctions.cpp:110:153: error: no matching function for call to ‘make_pair(SPL::list<SPL::rstring>&, curl_slist*&)’
             std::pair<HeaderCache::iterator,bool> res = headerCache->insert(std::make_pair<SPL::list<SPL::rstring>,struct curl_slist *>(theList,toReturn));
                                                                                                                                                         ^
src/httpFunctions.cpp:110:153: note: candidate is:
In file included from /usr/include/c++/4.8.2/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/4.8.2/bits/char_traits.h:39,
                 from /usr/include/c++/4.8.2/ios:40,
                 from /usr/include/c++/4.8.2/ostream:38,
                 from /usr/include/c++/4.8.2/iostream:39,
                 from /opt/ibm/InfoSphere_Streams/4.2.3.0/system/impl/include/SPL/Runtime/Type/TypeHelper.h:17,
                 from /opt/ibm/InfoSphere_Streams/4.2.3.0/include/SPL/Runtime/Type/Blob.h:24,
                 from ./include/httpFunctions.h:4,
                 from src/httpFunctions.cpp:1:
/usr/include/c++/4.8.2/bits/stl_pair.h:276:5: note: template<class _T1, class _T2> constexpr std::pair<typename std::__decay_and_strip<_Tp>::__type, typename std::__decay_and_strip<_T2>::__type> std::make_pair(_T1&&, _T2&&)
     make_pair(_T1&& __x, _T2&& __y)
     ^
/usr/include/c++/4.8.2/bits/stl_pair.h:276:5: note:   template argument deduction/substitution failed:
src/httpFunctions.cpp:110:153: note:   cannot convert ‘theList’ (type ‘SPL::list<SPL::rstring>’) to type ‘SPL::list<SPL::rstring>&&’
             std::pair<HeaderCache::iterator,bool> res = headerCache->insert(std::make_pair<SPL::list<SPL::rstring>,struct curl_slist *>(theList,toReturn));
                                                                                                                                              
Now the older c++ libs are broken.

joergboe commented:
Replace line 96 in the Makefile with:
$(OBJ)/httpClientAsync.o: $(SRC2)/httpClientAsync.cpp $(INCLUDES2)

joergboe commented:
I wonder why we download the release 0.13 of the cpp-netlib.
The latest release on github is 0.12.final
We should use stable releases of libs in out toolkits

joergboe commented:
The proposed way to build the cpp-netlib does not work:
~/git4/streamsx.inet/com.ibm.streamsx.inet/opt/cpp-netlib/cpp-netlib-0.13-release/build]$ cmake -DBOOST_ROOT=${BOOST_HOME}                                -DCPP-NETLIB_BUILD_SHARED_LIBS=ON                                -DCPP-NETLIB_BUILD_EXAMPLES=OFF                                -DCPP-NETLIB_BUILD_TESTS=OFF                                -DCPP-NETLIB_STATIC_BOOST=ON                                -DCMAKE_BUILD_TYPE=Release                                -DCMAKE_POSITION_INDEPENDENT_CODE=ON                                -DCMAKE_SHARED_LIBRARY_SONAME_CXX_FLAG='-U'                                -DCMAKE_SHARED_LINKER_FLAGS='-Wl,--whole-archive -lboost_system -lboost_thread -Wl,--no-whole-archive' ..
-- Linking boost testing libs dynamically...
CMake Error at /usr/share/cmake/Modules/FindBoost.cmake:1138 (message):
  Unable to find the requested Boost libraries.

  Boost version: 1.64.0

  Boost include path:
  /home/joergboe/git4/streamsx.inet/com.ibm.streamsx.inet/opt/downloaded/boost_1_64_0


  Could not find the following static Boost libraries:

          boost_system
          boost_thread

  No Boost libraries were found.  You may need to set BOOST_LIBRARYDIR to the
  directory containing Boost libraries or BOOST_ROOT to the location of
  Boost.
Call Stack (most recent call first):
  CMakeLists.txt:53 (find_package)


CMake Error at CMakeLists.txt:148 (export):
  export given target "cppnetlib-client-connections" which is not built by
  this project.


-- Configuring incomplete, errors occurred!

leongor commented:
> No Boost libraries were found. You may need to set BOOST_LIBRARYDIR to the
directory containing Boost libraries or BOOST_ROOT to the location of
Boost.

Please, follow all previous steps from the instructions:

> pushd boost_1_64_0 && BOOST_HOME=`pwd`

joergboe commented:
I got it

leongor commented:
@joergboe 

> I wonder why we download the release 0.13 of the cpp-netlib.

Good question. Today v0.13 is rc0. I chose it because of 2 reasons:

1. cpp-netlib uses internally asio library that comes with dual license: or as part of the boost library or as a standalone library. cpp-netlib used boost.asio till v0.12 where they decided to switch to a standalone (meaning adding the whole asio to the toolkit). In v0.13 they switched back to the boost.asio - means v0.12 is not so good for us.
2. Bug fixes.

joergboe commented:
Getting more issues:
...failed gcc.compile.c++ bin.v2/libs/python/build/gcc-4.8.5/release/link-static/threading-multi/numpy/dtype.o...
gcc.compile.c++ bin.v2/libs/python/build/gcc-4.8.5/release/link-static/threading-multi/numpy/matrix.o
In file included from ./boost/python/detail/prefix.hpp:13:0,
                 from ./boost/python/args.hpp:8,
                 from ./boost/python.hpp:11,
                 from ./boost/python/numpy/internal.hpp:17,
                 from libs/python/src/numpy/matrix.cpp:8:
./boost/python/detail/wrap_python.hpp:50:23: fatal error: pyconfig.h: No such file or directory
 # include <pyconfig.h>
                       ^
compilation terminated.

...failed updating 35 targets...
...skipped 9 targets...
...updated 623 targets...

When python3 is in PATH

joergboe commented:
When I use the standard python 2.7.5 I got other issue:
gcc.compile.c++ bin.v2/libs/fiber/build/gcc-4.8.5/release/link-static/threading-multi/context.o
libs/fiber/src/context.cpp: In constructor ‘boost::fibers::context::context(boost::fibers::dispatcher_context_t, const boost::context::preallocated&, const default_stack&, boost::fibers::scheduler*)’:
libs/fiber/src/context.cpp:236:14: error: call of overloaded ‘callcc(const std::allocator_arg_t&, const boost::context::preallocated&, const default_stack&, boost::fibers::context::context(boost::fibers::dispatcher_context_t, const boost::context::preallocated&, const default_stack&, boost::fibers::scheduler*)::__lambda5)’ is ambiguous
             });
              ^
libs/fiber/src/context.cpp:236:14: note: candidates are:
In file included from ./boost/fiber/context.hpp:28:0,
                 from libs/fiber/src/context.cpp:7:
./boost/context/continuation.hpp:469:1: note: boost::context::continuation boost::context::callcc(std::allocator_arg_t, StackAlloc, Fn&&, Arg ...) [with StackAlloc = boost::context::preallocated; Fn = const boost::context::basic_fixedsize_stack<boost::context::stack_traits>&; Arg = {boost::fibers::context::context(boost::fibers::dispatcher_context_t, const boost::context::preallocated&, const default_stack&, boost::fibers::scheduler*)::__lambda5}]
 callcc( std::allocator_arg_t, StackAlloc salloc, Fn && fn, Arg ... arg) {
 ^
./boost/context/continuation.hpp:483:1: note: boost::context::continuation boost::context::callcc(std::allocator_arg_t, boost::context::preallocated, StackAlloc, Fn&&, Arg ...) [with StackAlloc = boost::context::basic_fixedsize_stack<boost::context::stack_traits>; Fn = boost::fibers::context::context(boost::fibers::dispatcher_context_t, const boost::context::preallocated&, const default_stack&, boost::fibers::scheduler*)::__lambda5; Arg = {}]
 callcc( std::allocator_arg_t, preallocated palloc, StackAlloc salloc, Fn && fn, Arg ... arg) {
 ^
./boost/context/continuation.hpp:514:1: note: boost::context::continuation boost::context::callcc(std::allocator_arg_t, boost::context::preallocated, StackAlloc, Fn&&) [with StackAlloc = boost::context::basic_fixedsize_stack<boost::context::stack_traits>; Fn = boost::fibers::context::context(boost::fibers::dispatcher_context_t, const boost::context::preallocated&, const default_stack&, boost::fibers::scheduler*)::__lambda5]
 callcc( std::allocator_arg_t, preallocated palloc, StackAlloc salloc, Fn && fn) {
 ^
./boost/context/continuation.hpp:457:1: note: boost::context::continuation boost::context::callcc(Fn&&, Arg ...) [with Fn = const std::allocator_arg_t&; Arg = {boost::context::preallocated, boost::context::basic_fixedsize_stack<boost::context::stack_traits>, boost::fibers::context::context(boost::fibers::dispatcher_context_t, const boost::context::preallocated&, const default_stack&, boost::fibers::scheduler*)::__lambda5}; <template-parameter-1-3> = void]
 callcc( Fn && fn, Arg ... arg) {
 ^

    "g++"  -ftemplate-depth-128 -O3 -finline-functions -Wno-inline -Wall -pthread -m64 -std=c++11 -fPIC -DBOOST_ALL_NO_LIB=1 -DBOOST_DISABLE_ASSERTS -DBOOST_FIBERS_SOURCE -DNDEBUG  -I"." -c -o "bin.v2/libs/fiber/build/gcc-4.8.5/release/link-static/threading-multi/context.o" "libs/fiber/src/context.cpp"

...failed gcc.compile.c++ bin.v2/libs/fiber/build/gcc-4.8.5/release/link-static/threading-multi/context.o...

leongor commented:
@joergboe 
I think we can solve it by limiting the boost build only to 'system' and 'thread' libraries:
`scl enable devtoolset-4 "./b2 -j8 toolset=gcc variant=release link=static threading=multi address-model=64 cxxflags='-std=c++11 -fPIC' --with-system --with-thread"`
It'll also benefit the compile time significantly.

joergboe commented:
I can not recommend to use this library.
We must keep in mind that the inet toolkit is a generic toolkit an must support all platforms which are relevant for Streams. 
There are too much issues to build the lib and there are too much requirements through the cpp-netlib. This multiplied with the 5 platforms we support produces too much maintenance effort for this operator.
Especially there is a replacement for the cpp-netlib.
The curl lib is widely used, well tested and is available in all platforms without any additional effort.
The curl lib is already used in the inet toolkit.
I recommend to change the code and replace the cpp-netlib with libcurl

leongor commented:
@joergboe I understand your concerns, but as mentioned above libcurl is not async/non blocking.
I'm going to open request for the new httpasync toolkit.

chanskw commented:
@leongor even with an httpasync toolkit, we would still need to have legal reviews and approvals before we can include it on Github.

leongor commented:
@chanskw Who says otherwise?

chanskw commented:
@leongor not sure i understand the question... are you asking if someone else can override this requirement?

leongor commented:
@chanskw No, I mean no one says otherwise. The library should be approved before it's accepted.

chanskw commented:
ah ok!  Thank you!

joergboe commented:
> I understand your concerns, but as mentioned above libcurl is not async/non blocking. I'm going to open request for the new httpasync toolkit.
@leongor Thats why there is the common pattern to spawn the own threads in the process(int) method:
`void MY_OPERATOR::allPortsReady() {
    // Notifies that all ports are ready. No tuples should be submitted before
    // this. Source operators can use this method to spawn threads.
    createThreads(1); // Create source thread
  ....
}`

joergboe commented:
an alternative approach is in PR #310

leongor commented:
@joergboe 

> Thats why there is the common pattern to spawn the own threads in the process(int) method

Nothing wrong with opening user threads from inside the library. There a lot of operators that do exactly that - create 1 operator thread (manageable by the operator) that guards application threads (unmanaged by the operator). Here the explanation from the [knowledge center](https://www.ibm.com/support/knowledgecenter/SSCRJU_4.2.1/com.ibm.streams.dev.doc/doc/implementingoperators.html):

>   The SPL language run time considers the execution of a source operator complete when all threads of the operator complete their execution. Since source operators do not have tuple or punctuation processing functions, the completion of the operator threads marks the end of the execution of the operator. The language run time ensures that final punctuation is sent on all output ports upon completion of the execution of the source operator. **The runtime keeps track only of threads that are created with the SPL runtime APIs. If the operator relies on other threads, such as those created by third-party libraries, then an SPL operator thread must be created to wait on the external threads.** Otherwise, the run time prematurely sends final punctuation on the output ports.

leongor commented:
> an alternative approach is in PR #310

What do you mean? This PR is #310.

joergboe commented:
As I learned in the last call, that the number of parallel connections is unlimited in this operator.
The operator need a control over the maximum number of parallel connections.
Otherwise it may happen that one operator consumes all system resources (open files/tcp ports) at one go. This must be prevented. Thus the control over the number of connections is required.

leongor commented:
I'll investigate if it's possible to apply this setting as a parameter on the operator level. Otherwise a threaded port can easily limit a number of tuples that handled concurrently (thus a number of TCP connections accordingly).

joergboe commented:
I would like to open an separate directory for those experimental operators.
This directory should contain a full toolkit structure an can be handled from an spl application as a separate toolkit.

chanskw commented:
Instead of directory, perhaps set up a feature branch?

joergboe commented:
@chanskw I would like to have it in a separate directory which sets up a separate toolkit that contains operators with experimental operators which are more likely changed.
But these operators share the namespace with the inet toolit.

leongor commented:
As agreed, I removed HTTPRequestAsync operator from com.ibm.streamsx.inet toolkit and added it as a separate com.ibm.streamsx.inet.async toolkit.
Few important points:
1. AWSOM was approved for cpp-netlib library.
2. As @joergboe requested HTTPRequestAsync has a new connectionCap parameter that allows to limit a number of open client connections.
3. Automatic SNI and chunks handling was added.

joergboe commented on .gitignore:
```diff
@@ -13,6 +13,7 @@ toolkit.xml
 /com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/*/*.xml
 /com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/*/*.gif
 !com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/native.function/function.xml
+!com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/HTTPRequestAsync/HTTPRequestAsync.xml
```
Why do need this exception for  HTTPRequestAsync ?

joergboe commented on .gitignore:
```diff
@@ -13,6 +13,7 @@ toolkit.xml
 /com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/*/*.xml
 /com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/*/*.gif
 !com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/native.function/function.xml
+!com.ibm.streamsx.inet/com.ibm.streamsx.inet.http/HTTPRequestAsync/HTTPRequestAsync.xml
```
In general used librarys should be downloaded during build process and should not be copied into the repository. 
should do something with wget or similar to download during build and add a link to the licence or copy the licence.

leongor replied:
This is for C++ operator model, not for libraries (as you remember, you suggested to add it because as opposed to Java operator model being generated, the C++ operator model is persistent).
For libraries there is an `ant` file to download and build them.

joergboe commented:
In the sample HTTPClientAsync the is the SPL Build broken.
The sample uses the Studion build, but there is no file .settings/XXX-BuildConfig.splbuild

You should either complete the spl build configuration
or 
add a make-file (like the other samples have) and swithc the stuio settings to 'external build'

leongor commented:
Added external build with makefile.

joergboe commented:
The code is published in branch HTTPRequestAsync.
